src/doc_table.c

```diff
@@ -39,24 +39,34 @@ static inline int DocTable_ValidateDocId(const DocTable *t, t_docId docId) {
   return docId != 0 && docId <= t->maxDocId;
 }
 
-RSDocumentMetadata *DocTable_Get(const DocTable *t, t_docId docId) {
+const RSDocumentMetadata *DocTable_Get(const DocTable *t, t_docId docId) {
   if (!DocTable_ValidateDocId(t, docId)) {
     return NULL;
   }
   uint32_t bucketIndex = DocTable_GetBucket(t, docId);
   if (bucketIndex >= t->cap) {
     return NULL;
   }
+  // While we iterate over the chain, we have locked the index spec (R/W), so we either a writer alone or
+  // multiple readers. In any case, we can safely iterate over the chain without a lock and
+  // increment the ref count of the document metadata when we find it.
   DMDChain *dmdChain = &t->buckets[bucketIndex];
   DLLIST2_FOREACH(it, &dmdChain->lroot) {
     RSDocumentMetadata *dmd = DLLIST2_ITEM(it, RSDocumentMetadata, llnode);
     if (dmd->id == docId) {
-      return dmd;
+      return DMD_Incref(dmd);
     }
   }
   return NULL;
 }
 
+static RSDocumentMetadata *DocTable_GetOwn(const DocTable *t, t_docId docId) {
```

Comment:
consider doing the other way
`DocTable_Get` should call `DocTable_GetOwn` and increase the refcount rather than increase and decrease